File: Src/Compilers/CSharp/Portable/Declarations/SingleTypeDeclaration.cs
// Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Diagnostics;
using Microsoft.CodeAnalysis.CSharp.Symbols;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Text;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.CSharp
{
    internal sealed class SingleTypeDeclaration : SingleNamespaceOrTypeDeclaration
    {
        private readonly DeclarationKind kind;
        private readonly TypeDeclarationFlags flags;
        private readonly ushort arity;
        private readonly DeclarationModifiers modifiers;
        private readonly ImmutableArray<SingleTypeDeclaration> children;
        private readonly ICollection<string> memberNames;

        [Flags()]
        internal enum TypeDeclarationFlags : byte
        {
            None = 0,
            AnyMemberHasExtensionMethodSyntax = 1 << 1,
            HasAnyAttributes = 1 << 2,
            HasBaseDeclarations = 1 << 3,
            AnyMemberHasAttributes = 1 << 4,
            HasAnyNontypeMembers = 1 << 5,
            HasPrimaryCtor = 1 << 6,
        }

        internal SingleTypeDeclaration(
            DeclarationKind kind,
            string name,
            int arity,
            DeclarationModifiers modifiers,
            TypeDeclarationFlags declFlags,
            SyntaxReference syntaxReference,
            SourceLocation nameLocation,
            ICollection<string> memberNames,
            ImmutableArray<SingleTypeDeclaration> children)
            : base(name,
                   syntaxReference,
                   nameLocation)
        {
            Debug.Assert(kind != DeclarationKind.Namespace);

            this.kind = kind;
            this.arity = (ushort)arity;
            this.modifiers = modifiers;
            this.memberNames = memberNames;
            this.children = children;
            this.flags = declFlags;
        }

        public override DeclarationKind Kind
        {
            get
            {
                return this.kind;
            }
        }

        public new ImmutableArray<SingleTypeDeclaration> Children
        {
            get
            {
                return children;
            }
        }

        public int Arity
        {
            get
            {
                return this.arity;
            }
        }

        public DeclarationModifiers Modifiers
        {
            get
            {
                return this.modifiers;
            }
        }

        public ICollection<string> MemberNames
        {
            get
            {
                return memberNames;
            }
        }

        public bool AnyMemberHasExtensionMethodSyntax
        {
            get
            {
                return (this.flags & TypeDeclarationFlags.AnyMemberHasExtensionMethodSyntax) != 0;
            }
        }

        public bool HasAnyAttributes
        {
            get
            {
                return (this.flags & TypeDeclarationFlags.HasAnyAttributes) != 0;
            }
        }

        public bool HasBaseDeclarations
        {
            get
            {
                return (this.flags & TypeDeclarationFlags.HasBaseDeclarations) != 0;
            }
        }

        public bool AnyMemberHasAttributes
        {
            get
            {
                return (this.flags & TypeDeclarationFlags.AnyMemberHasAttributes) != 0;
            }
        }

        public bool HasAnyNontypeMembers
        {
            get
            {
                return (this.flags & TypeDeclarationFlags.HasAnyNontypeMembers) != 0;
            }
        }

        public bool HasPrimaryCtor
        {
            get
            {
                return (this.flags & TypeDeclarationFlags.HasPrimaryCtor) != 0;
            }
        }

        protected override ImmutableArray<SingleNamespaceOrTypeDeclaration> GetNamespaceOrTypeDeclarationChildren()
        {
            return StaticCast<SingleNamespaceOrTypeDeclaration>.From(children);
        }

        internal TypeDeclarationIdentity Identity
        {
            get
            {
                return new TypeDeclarationIdentity(this);
            }
        }

        // identity that is used when collecting all declarations 
        // of same type across multiple containers
        internal struct TypeDeclarationIdentity : IEquatable<TypeDeclarationIdentity>
        {
            private readonly SingleTypeDeclaration decl;

            internal TypeDeclarationIdentity(SingleTypeDeclaration decl)
            {
                this.decl = decl;
            }

            public override bool Equals(object obj)
            {
                return obj is TypeDeclarationIdentity && Equals((TypeDeclarationIdentity)obj);
            }

            public bool Equals(TypeDeclarationIdentity other)
            {
                var thisDecl = this.decl;
                var otherDecl = other.decl;

                // same as itself
                if ((object)thisDecl == otherDecl)
                {
                    return true;
                }

                // arity, kind, name must match
                if ((thisDecl.arity != otherDecl.arity) ||
                    (thisDecl.kind != otherDecl.kind) ||
                    (thisDecl.name != otherDecl.name))
                {
                    return false;
                }

                if (thisDecl.kind == DeclarationKind.Enum || thisDecl.kind == DeclarationKind.Delegate)
                {
                    // oh, so close, but enums and delegates cannot be partial
                    return false;
                }

                return true;
            }

            public override int GetHashCode()
            {
                var thisDecl = this.decl;
                return Hash.Combine(thisDecl.Name.GetHashCode(),
                    Hash.Combine(thisDecl.Arity.GetHashCode(),
                    (int)thisDecl.Kind));
            }
        }
    }
}

File: src/Compilers/VisualBasic/Portable/Declarations/SingleTypeDeclaration.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports System.Collections.Immutable
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols

Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
    Friend Class SingleTypeDeclaration
        Inherits SingleNamespaceOrTypeDeclaration

        Private ReadOnly _children As ImmutableArray(Of SingleTypeDeclaration)

        ' CONSIDER: It may be possible to merge the flags and arity into a single
        ' 32-bit quantity if space is needed.

        Private ReadOnly _kind As DeclarationKind
        Private ReadOnly _flags As TypeDeclarationFlags
        Private ReadOnly _arity As UShort
        Private ReadOnly _modifiers As DeclarationModifiers

        Public ReadOnly QuickAttributes As QuickAttributes

        Friend Enum TypeDeclarationFlags As Byte
            None = 0
            HasAnyAttributes = 1 << 1
            HasBaseDeclarations = 1 << 2
            AnyMemberHasAttributes = 1 << 3
            HasAnyNontypeMembers = 1 << 4
        End Enum

        Public Sub New(kind As DeclarationKind,
                       name As String,
                       arity As Integer,
                       modifiers As DeclarationModifiers,
                       declFlags As TypeDeclarationFlags,
                       syntaxReference As SyntaxReference,
                       nameLocation As Location,
                       memberNames As ImmutableHashSet(Of String),
                       children As ImmutableArray(Of SingleTypeDeclaration),
                       quickAttributes As QuickAttributes)
            MyBase.New(name, syntaxReference, nameLocation)

            Debug.Assert(kind <> DeclarationKind.Namespace)

            Me._kind = kind
            Me._arity = CUShort(arity)
            Me._flags = declFlags
            Me._modifiers = modifiers
            Me.MemberNames = memberNames
            Me._children = children
            Me.QuickAttributes = quickAttributes
        End Sub

        Public Overrides ReadOnly Property Kind As DeclarationKind
            Get
                Return Me._kind
            End Get
        End Property

        Public ReadOnly Property Arity As Integer
            Get
                Return _arity
            End Get
        End Property

        Public ReadOnly Property HasAnyAttributes As Boolean
            Get
                Return (_flags And TypeDeclarationFlags.HasAnyAttributes) <> 0
            End Get
        End Property

        Public ReadOnly Property HasBaseDeclarations As Boolean
            Get
                Return (_flags And TypeDeclarationFlags.HasBaseDeclarations) <> 0
            End Get
        End Property

        Public ReadOnly Property HasAnyNontypeMembers As Boolean
            Get
                Return (_flags And TypeDeclarationFlags.HasAnyNontypeMembers) <> 0
            End Get
        End Property

        Public ReadOnly Property AnyMemberHasAttributes As Boolean
            Get
                Return (_flags And TypeDeclarationFlags.AnyMemberHasAttributes) <> 0
            End Get
        End Property

        Public Overloads ReadOnly Property Children As ImmutableArray(Of SingleTypeDeclaration)
            Get
                Return _children
            End Get
        End Property

        Public ReadOnly Property Modifiers As DeclarationModifiers
            Get
                Return Me._modifiers
            End Get
        End Property

        Public ReadOnly Property MemberNames As ImmutableHashSet(Of String)

        Protected Overrides Function GetNamespaceOrTypeDeclarationChildren() As ImmutableArray(Of SingleNamespaceOrTypeDeclaration)
            Return StaticCast(Of SingleNamespaceOrTypeDeclaration).From(_children)
        End Function

        Private Function GetEmbeddedSymbolKind() As EmbeddedSymbolKind
            Return SyntaxReference.SyntaxTree.GetEmbeddedKind()
        End Function

        Private NotInheritable Class Comparer
            Implements IEqualityComparer(Of SingleTypeDeclaration)

            Private Shadows Function Equals(decl1 As SingleTypeDeclaration, decl2 As SingleTypeDeclaration) As Boolean Implements IEqualityComparer(Of SingleTypeDeclaration).Equals
                ' We should not merge types across embedded code / user code boundary.
                Return IdentifierComparison.Equals(decl1.Name, decl2.Name) _
                        AndAlso decl1.Kind = decl2.Kind _
                        AndAlso decl1.Kind <> DeclarationKind.Enum _
                        AndAlso decl1.Arity = decl2.Arity _
                        AndAlso decl1.GetEmbeddedSymbolKind() = decl2.GetEmbeddedSymbolKind()
            End Function

            Private Shadows Function GetHashCode(decl1 As SingleTypeDeclaration) As Integer Implements IEqualityComparer(Of SingleTypeDeclaration).GetHashCode
                Return Hash.Combine(IdentifierComparison.GetHashCode(decl1.Name), Hash.Combine(decl1.Arity.GetHashCode(), CType(decl1.Kind, Integer)))
            End Function
        End Class

        Public Shared ReadOnly EqualityComparer As IEqualityComparer(Of SingleTypeDeclaration) = New Comparer()
    End Class
End Namespace
